File: ios/pdfme_flutter/Sources/pdfme_flutter/PdfmeFlutterPlugin.swift
import Flutter
import UIKit

public class PdfmeFlutterPlugin: NSObject, FlutterPlugin {
  private let engine = PdfmeEngine()
  private let serial = DispatchQueue(label: "com.pdfme.pdfme_flutter.engine")

  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "pdfme_flutter", binaryMessenger: registrar.messenger())
    let instance = PdfmeFlutterPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "initialize":
      Task {
        do {
          try await engine.initialize()
          DispatchQueue.main.async { result(nil) }
        } catch let error as PdfmeNativeError {
          DispatchQueue.main.async { result(FlutterError(code: error.code, message: error.message, details: nil)) }
        } catch {
          DispatchQueue.main.async {
            result(FlutterError(code: "RUNTIME_INIT_ERROR", message: error.localizedDescription, details: nil))
          }
        }
      }
    case "generate":
      guard
        let args = call.arguments as? [String: Any],
        let templateJson = args["templateJson"] as? String,
        let inputsJson = args["inputsJson"] as? String
      else {
        result(FlutterError(code: "INVALID_INPUT", message: "templateJson and inputsJson are required", details: nil))
        return
      }
      let optionsJson = (args["optionsJson"] as? String) ?? "{}"
      Task {
        do {
          let data = try await engine.generate(
            templateJson: templateJson,
            inputsJson: inputsJson,
            optionsJson: optionsJson
          )
          DispatchQueue.main.async { result(data) }
        } catch let error as PdfmeNativeError {
          DispatchQueue.main.async { result(FlutterError(code: error.code, message: error.message, details: nil)) }
        } catch {
          DispatchQueue.main.async {
            result(FlutterError(code: "GENERATION_ERROR", message: error.localizedDescription, details: nil))
          }
        }
      }
    case "dispose":
      engine.dispose()
      result(nil)
    case "getPlatformVersion":
      result("iOS " + UIDevice.current.systemVersion)
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}

File: android/src/main/kotlin/com/pdfme/pdfme_flutter/PdfmeFlutterPlugin.kt
package com.pdfme.pdfme_flutter

import android.os.Handler
import android.os.Looper
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.concurrent.Executors

/** PdfmeFlutterPlugin — bridges Dart to the offscreen pdfme JS engine. */
class PdfmeFlutterPlugin :
    FlutterPlugin,
    MethodCallHandler {
    private lateinit var channel: MethodChannel
    private var engine: PdfmeEngine? = null
    private val executor = Executors.newSingleThreadExecutor()
    private val mainHandler = Handler(Looper.getMainLooper())

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "pdfme_flutter")
        channel.setMethodCallHandler(this)
        engine = PdfmeEngine(flutterPluginBinding.applicationContext)
    }

    override fun onMethodCall(
        call: MethodCall,
        result: Result,
    ) {
        when (call.method) {
            "initialize" -> {
                executor.execute {
                    try {
                        engine?.initialize()
                            ?: throw PdfmeNativeException("RUNTIME_INIT_ERROR", "Engine missing")
                        mainHandler.post { result.success(null) }
                    } catch (e: PdfmeNativeException) {
                        mainHandler.post { result.error(e.code, e.message, null) }
                    } catch (t: Throwable) {
                        mainHandler.post {
                            result.error("RUNTIME_INIT_ERROR", t.message, null)
                        }
                    }
                }
            }
            "generate" -> {
                val templateJson = call.argument<String>("templateJson")
                val inputsJson = call.argument<String>("inputsJson")
                val optionsJson = call.argument<String>("optionsJson") ?: "{}"
                if (templateJson.isNullOrEmpty() || inputsJson.isNullOrEmpty()) {
                    result.error("INVALID_INPUT", "templateJson and inputsJson are required", null)
                    return
                }
                executor.execute {
                    try {
                        val bytes =
                            engine?.generate(templateJson, inputsJson, optionsJson)
                                ?: throw PdfmeNativeException("RUNTIME_INIT_ERROR", "Engine missing")
                        mainHandler.post { result.success(bytes) }
                    } catch (e: PdfmeNativeException) {
                        mainHandler.post { result.error(e.code, e.message, null) }
                    } catch (t: Throwable) {
                        mainHandler.post {
                            result.error("GENERATION_ERROR", t.message, null)
                        }
                    }
                }
            }
            "dispose" -> {
                executor.execute {
                    try {
                        engine?.dispose()
                        mainHandler.post { result.success(null) }
                    } catch (t: Throwable) {
                        mainHandler.post {
                            result.error("RUNTIME_INIT_ERROR", t.message, null)
                        }
                    }
                }
            }
            "getPlatformVersion" -> result.success("Android ${android.os.Build.VERSION.RELEASE}")
            else -> result.notImplemented()
        }
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        executor.execute {
            engine?.dispose()
            engine = null
        }
        executor.shutdown()
    }
}
